### PR TITLE
Suggested fixes for `pub` and `priv` parser errors

### PR DESCRIPTION
**Pull request by dsainati1**
Closes https://github.com/onflow/cadence/issues/2741

## Description

Add suggested fixes to parser errors for old `pub` and `priv` keywords, and refactors `HasSuggestedFixes` to make this possible

______

- [ ] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [X] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

File: runtime/parser/errors.go [hunk 1]
```diff
@@ -91,6 +91,60 @@ func (e *SyntaxError) Error() string {
 	return e.Message
 }
 
+// SyntaxErrorWithSuggestedFix
+
+type SyntaxErrorWithSuggestedFix struct {
```
Comment (turbolent) [inline runtime/parser/errors.go:96]: Maybe rename to `SyntaxErrorWithSuggestReplacement` to indicate that the fix is a replacement, and not a removal or insertion.

File: runtime/parser/errors.go [hunk 2]
```diff
@@ -91,6 +91,60 @@ func (e *SyntaxError) Error() string {
...
+	Message      string
+	SuggestedFix string
+	Pos          ast.Position
```
Comment (turbolent) [inline runtime/parser/errors.go:99]: This probably needs a whole range, see below

File: runtime/parser/errors.go [hunk 3]
```diff
@@ -91,6 +91,60 @@ func (e *SyntaxError) Error() string {
...
+	return []errors.SuggestedFix[ast.TextEdit]{
+		{
+			Message: fmt.Sprintf("replace with %s", e.SuggestedFix),
+			TextEdits: []ast.TextEdit{
+				{
+					Replacement: e.SuggestedFix,
+					Range: ast.Range{
+						StartPos: e.Pos,
+						EndPos:   e.Pos,
+					},
```
Comment (turbolent) [inline runtime/parser/errors.go:141]: Not sure, but isn't this currently suggesting to replace "nothing" (start and and position are the same),so is effectively suggesting an insert? Maybe replace the single position (`Pos`) with a range.

**Review by turbolent — COMMENTED** [review]
Nice!

File: runtime/errors/errors.go [hunk 4]
```diff
@@ -114,6 +114,17 @@ type MemoryError struct {
...
 
+// SuggestedFix
+
+type HasSuggestedFixes[T any] interface {
+	SuggestFixes(code string) []SuggestedFix[T]
+}
+
+type SuggestedFix[T any] struct {
+	Message   string
+	TextEdits []T
```
[the remark this answers is not in the document]
Comment (dsainati1) [reply runtime/errors/errors.go:125]: `common/errors` cannot import `ast`, so I made this parameterizable over any tim

[the remark this answers is not in the document]
Comment (SupunS) [reply runtime/errors/errors.go:125]: Moving `TextEdit` to `runtime/errors` might be a solution?

**Review by SupunS — APPROVED** [review]

File: runtime/errors/errors.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dsainati1) [reply runtime/errors/errors.go:125]: `TextEdit` contains an `ast.Range`, which is the issue.

[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/errors/errors.go:125]: Ah, I see, that makes sense

**Review by turbolent — APPROVED** [review]
Nice!

**Pull request merged**